### PR TITLE
#2 mind builder

### PR DESCRIPTION
**Pull request by MMCQ33**
Broken UI, and needs added save button. But minds can be build and edited.

### CODE REVIEW

File: Hivemind/Assets/Scripts/Ant.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                    minds[i].Update(mindGroupMind[i]);
+                    if (!minds[i].Equals(mindGroupMind[i]))
+                    {
+                        minds[i] = mindGroupMind[i].Clone();
+                    }
+                }
+            }
+        }
+
+        double likeliest= 0;
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Ant.cs]: Spacing between =
Recommend some automated cleanup on all changed document

File: Hivemind/Assets/Scripts/MindGroup.cs [hunk 2]
```diff
@@ -15,9 +15,25 @@ public class MindGroup
...
-    public MindGroup(GameObject UiObject)
+    public IMind Mind { get; private set; } = new Gathering(ResourceType.Unknown, 1, Gathering.Direction.None);
+
+    public List<IMind> minds { get; private set; }
+
+    public CombatMind combatMind { get; private set; }
+    
+    public int mindPoints { get; set; }
+
+    public MindGroup(GameObject UiObject, IMind mind = null)
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/MindGroup.cs]: mind in constructor is not used

File: Hivemind/Assets/Scripts/MindGroup.cs [hunk 3]
```diff
@@ -15,9 +15,25 @@ public class MindGroup
 
     private List<UnitGroup> unitList;
 
-    public MindGroup(GameObject UiObject)
+    public IMind Mind { get; private set; } = new Gathering(ResourceType.Unknown, 1, Gathering.Direction.None);
+
+    public List<IMind> minds { get; private set; }
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/MindGroup.cs]: Public variables should start with a capital letter

File: Hivemind/Assets/Scripts/Minds/CombatFight.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Minds/CombatFight.cs]: Unused using, recommend automated cleanup for all files

File: Hivemind/Assets/Scripts/Minds/CombatMind.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
+﻿
+public class CombatMind : IMind
+{
+    private float minEstimetedDifference;
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Minds/CombatMind.cs]: Estimated

Comment (ErikLouwerse) [reply Hivemind/Assets/Scripts/Minds/CombatMind.cs]: Use visual studios Rename so you don't have to manually change everything

File: Hivemind/Assets/Scripts/Minds/CombatMind.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    public void Execute(Ant ant)
+    {
+            if (this == null)
+            {
+                 new CombatFight().Execute(ant);
+            }
+
+            float healthPercantageDifference = ((float)ant.health / (float)ant.closestEnemy.health);
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Minds/CombatMind.cs]: Percentage

Comment (ErikLouwerse) [reply Hivemind/Assets/Scripts/Minds/CombatMind.cs]: Also to get the percentage you should do * 100

Comment (ErikLouwerse) [reply Hivemind/Assets/Scripts/Minds/CombatMind.cs]: And also -100, recommend removing the percentage word from this variable and the one below

File: Hivemind/Assets/Scripts/Minds/CombatMind.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            float damagePercantageDifference = ((float)ant.damage / (float)ant.closestEnemy.damage);
+            float strengthDifference = (healthPercantageDifference * 1 + damagePercantageDifference * 2) / 3;
+
+            if (strengthDifference >= minEstimetedDifference)
+            {
+                 new CombatFight().Execute(ant);
+            }
+             new CombatFlee().Execute(ant);
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Minds/CombatMind.cs]: Shouldn't this be in an else?

File: Hivemind/Assets/Scripts/Minds/Gathering.cs [hunk 8]
```diff
@@ -29,105 +30,116 @@ public enum Direction
...
+
     private Dictionary<ResourceType, int> inventory;
     private int nextHarvest;
-    private ResourceNode target;
-    private Storage storage;
-    private NavMeshAgent agent;
-    private float baseSpeed;
-    private List<GameObject> carryingObjects;
+
+    public List<GameObject> carryingObjects;
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Minds/Gathering.cs]: Variable is only used in this class, should revert to private

File: Hivemind/Assets/Scripts/Minds/Gathering.cs [hunk 9]
```diff
@@ -29,105 +30,116 @@ public enum Direction
...
-    private ResourceNode findResource()
+    public Gathering(ResourceType resType, int carryweight, Direction exploreDirection, bool isScout = true)
+    {
+        prefferedType = resType;
+        carryWeight = carryweight;
+        prefferedDirection = exploreDirection;
+        IsScout = isScout;
+    }
+
+    private ResourceNode findResource(ResourceType PrefferedResource)
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Minds/Gathering.cs]: Why give the PrefferedResource when you already have acces to it in the class?

File: Hivemind/Assets/Scripts/Minds/Gathering.cs [hunk 10]
```diff
@@ -190,26 +209,27 @@ private void Update()
...
+                            ant.currentSpeed = ant.baseSpeed;
+                            ant.UpdateSpeed();
                         }
-                        state = State.Idle;
+                        ant.state = State.Idle;
                     }
                 }
                 else
                 {
-                    storage = GameWorld.GetStorage();
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Minds/Gathering.cs]: What if storage cannot be found?

File: Hivemind/Assets/Scripts/UI/CombatMindUI.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+
+public class CombatMindUI : MonoBehaviour, IMindUI
+{
+    public GameObject Ui_IconObj { get; set ; }
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/UI/CombatMindUI.cs]: Remove Useless getset

File: Hivemind/Assets/Scripts/UI/IMindUI.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using UnityEngine;
+
+interface IMindUI
+{
+    GameObject Ui_IconObj { get; set; }
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/UI/IMindUI.cs:5]: remove getset

File: Hivemind/Assets/Scripts/UI/MindBuilderScript.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+{
+    private Vector2 lastMousePosition;
+
+    //currently dragged mind object.
+    private IMindUI mindTypeObj;
+
+    public MindGroup mindGroup;
+
+    // Start is called before the first frame update
+    void Start()
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/UI/MindBuilderScript.cs]: Remove unused function

File: Hivemind/Assets/Scripts/UI/ResourceMindUI.cs [hunk 14]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+
+
+
+public class ResourceMindUI : MonoBehaviour, IMindUI
+{
+    public GameObject Ui_IconObj { get; set; }
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/UI/ResourceMindUI.cs]: getset

File: Hivemind/Assets/Scripts/Ui/UiController.cs [hunk 15]
```diff
@@ -24,6 +24,8 @@ public class UiController : MonoBehaviour, IInitializePotentialDragHandler, IDra
     [SerializeField]
     private GameObject mindBuilderPanel;
 
+    private MindGroup CurrentOpenMindGroup;
```
Comment (ErikLouwerse) [inline Hivemind/Assets/Scripts/Ui/UiController.cs]: private field should not start with capital letter

**Review by ErikLouwerse — CHANGES_REQUESTED** [review]

**Review by ErikLouwerse — APPROVED** [review]

**Review by zoompower — APPROVED** [review]

**Pull request merged**